### PR TITLE
[Asana] refactoring and Oauth flow

### PR DESCRIPTION
**Pull request by EugeneLightsOn**
### What's being changed:

- Refactors client code into client.py
- Oauth flow
- Clarifies code and improve README

### How did you test this change (include any code snippets, API requests, screenshots, or gifs):
image

### CODE REVIEW

File: asana/README.md [hunk 1]
```diff
@@ -7,10 +7,98 @@ A connector to integrate Asana, the project management tool, to Cohere. Featurin
 The Asana connector only offers full-text search for Task objects.
 
 ## Configuration
+The Asana connector provides two authentication methods: [Personal Access Token](https://developers.asana.com/docs/personal-access-token) and [OAuth 2.0](https://developers.asana.com/docs/oauth)
 
-To use this connector, head to your Asana Developer Console. From here, you can create a new Personal Access Token. Use this value for the `ASANA_ACCESS_TOKEN` environment variable. Then, head to your Admin Console. The URL at the top of your web browser should look like `https://app.asana.com/admin/<workspace_gid>`, grab the `<workspace_gid>` value and use it for `ASANA_WORKSPACE_GID`.
+If you are using a personal access token, you need to set the following environment variables:
```
Comment (tianjing-li) [inline asana/README.md:12]: To clarify the authentication methods, could you add sub sections please, eg a 
```
### (Method 1) Personal Access Token

### (Method 2) OAuth 2.0
```

File: asana/README.md [hunk 2]
```diff
@@ -7,10 +7,98 @@ A connector to integrate Asana, the project management tool, to Cohere. Featurin
...
+
+Then, head to your Admin Console. The URL at the top of your web browser should look like `https://app.asana.com/admin/<workspace_gid>`, 
+grab the `<workspace_gid>` value and use it for `ASANA_WORKSPACE_GID`.
+
+## Optional Configuration
+```
+ASANA_TASK_PROPERTIES
+```
+This variable may contain a comma-separated list of task properties (e.g. ["name","notes"]).
+By default, the following task properties are returned:
```
Comment (tianjing-li) [inline asana/README.md:54]: Can you link the docs for the task properties here

**Review by tianjing-li — APPROVED** [review]
Looks good

File: asana/README.md (hunk 1 above)
Comment (EugeneLightsOn) [reply asana/README.md:13]: It's done. Thank you.

File: asana/README.md (hunk 2 above)
Comment (EugeneLightsOn) [reply asana/README.md:58]: Thanks. It's done

File: asana/provider/client.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        self.task_properties = task_properties or self.DEFAULT_TASK_PROPERTIES
+        self.access_token = access_token
+        self.workspace_gid = workspace_gid
+        self.search_limit = search_limit or self.DEFAULT_SEARCH_LIMIT
+        self.fields_mapping = fields_mapping
+        self.headers = {"Authorization": f"Bearer {self.access_token}"}
+
+    def add_task_url_to_results(self, results):
+        for result in results:
+            result["url"] = f"{self.BASE_URL}/0/0/{result['gid']}"
```
[the remark this answers is not in the document]
Comment (EugeneLightsOn) [reply asana/provider/client.py]: Thank you @walterbm-cohere.

**Pull request merged**